### ISSUE TITLE
What's the use of `for`—`then` constructs?

### ISSUE THREAD

**Issue by loveencounterflow** [jashkenas/coffeescript#3999]
There's an issue (https://github.com/michaelficarra/CoffeeScriptRedux/issues/317) over at the CS Redux project where the possibility of a `for`—`then` loop as shown below is remarked (as i understand it the implication being that this won't work in Redux, yet). Indeed, a little test with CS 1.9.3 shows that it compiles and runs well:

```coffee
coffee> for [ 0 .. 3 ] then console.log 'x'
x
x
x
x
[ undefined,
  undefined,
  undefined,
  undefined ]
coffee>
```
But, to quote myself: """That it is legal doesn't mean it's intentional, very useful or clear. The [...] part [after `then`] gets evaluated the appropriate number of times and the value of the expression is a list with the same number of `undefined` values. [...] It gets worse if you happen to place this as the last line into a function, because then implicit `return` insertion kicks in and will make that list the return value of the function, which is very probably not what you intended."""

Should this be in the language? Is it intentional, maybe because it follows from some systematic consideration? What could be a real-world use case for this construct?

**Comment by vendethiel** [comment]
It's not a special syntax, you can just use `then` to introduce a block on that line (implemented by @satyr IIRC).

this works:

```
if a then b
```

**Comment by lydell** [comment]
Are you guys talking about `then`? Because michaelficarra/CoffeeScriptRedux#317 is about the special `for` syntax where the `VARIABLE in` part is left out: `for [a..b]`.

I know for sure that it is documented, because I wrote that documentation (see #3970, commit 99ealc7a). It’s in the [Loops and Comprehensions](http://coffeescript.org/#loops) part. (Search for `for [` in that page to get to it quickly.)

Before documenting it, I also knew for sure that it was intended, because I once reported and fixed a bug for it: #3671. That issue contains lots of useful information on it, as well as some links if you’d like to know even more.

Finally, here’s a place where I’ve used it: https://github.com/akhodakivskiy/VimFx/blob/f3aa603a70b2b51d438fc887089c15732d6a2834/extension/lib/commands.coffee#L270. (That’s also where I got the inspiration for the documentation, and how I initially found #3970: I usually add `by 1` to all range loops that involve a variable, as an optimization, but I had to leave it out there because of #3970.)

> the value of the expression is a list with the same number of `undefined` values

`undefined` because `console.log` returns `undefined`.

> It gets worse if you happen to place this as the last line into a function, because then implicit return insertion kicks in and will make that list the return value of the function, which is very probably not what you intended

That’s true for any function ending with _any_ loop and has nothing to do with this.

**Comment by vendethiel** [comment]
Looks like I read the explanation too fast (probably because I expect that). Thanks for the great write-up @lydell.

**Comment by vendethiel** [comment]
Actually, there's no reason to even let this one open. It's even documented and al :-)!

**Comment by loveencounterflow** [comment]
I understood the post to be about the `then` part, not the left-out variable part. Turns out the latter is indeed documented, if a bit on the short side, but where can i see that `for`...`then` is legal at all? 

Further, when i do `42 for [ 1 .. 3 ]` on the REPL, i get `[ 42, 42, 42 ]` as printout, but when i do `console.log 42 for [ 1 .. 3 ]` in a script (or on the REPL) i get `42` printed out three times. This sure is well within what is intended to be the case in CS, but I must say that if it surprises me after years of intense working with the language (to the point that i have to try everything myself to see what the results are), chances a clarification in the docs could really be helpful.

Also, `console.log ( for [ 0 ... 3 ] then 108 )` and `console.log ( 108 for [ 0 ... 3 ] )` seem to be equivalent. Maybe it would be good to point this out somewhere to clear up things.

**Comment by vendethiel** [comment]
> but I must say that if it surprises me after years of intense working with the language

I don't get what's surprising?

**Comment by lydell** [comment]
In case it helps, here is an example on how I've used `for ... then`: https://github.com/lydell/jawa-juice/blob/62a851d95f472214478e2ad037eb38b513adfd46/index.coffee#L55-69. That also demonstrates how to avoid returning the value of the loop.

**Comment by loveencounterflow** [comment]
@vendethiel the surprising thing here is actually more in the precedence of the assignment operator (i.e. one might expect that what gets echoes in the REPL is what gets assigned when prefixing the whole thing with a `x = `, but that is not the case here). 

Basically what i want to say with this example and the ones that opened this issue is not so much that this behavior is necessarily good or bad or useful or useless, my point is that the documentation does not mention some points which are radically different in CS when compared to other languages.

@lydell your example is very good. Am I wrong in saying that a clarification in the docs like "instead of indenting an expression to be looped over, it is also possible to introduce it with `then`" would be great?

**Comment by epidemian** [comment]
[The docs](http://coffeescript.org/#operators) make a brief comment about `then`:

> Instead of a newline or semicolon, `then` can be used to separate conditions from expressions, in **while**, **if/else**, and **switch/when** statements.

But they make no mention of `for`, probably because when used in a `for` construct `then` doesn't separate a *condition* from an expression. Also, the mention of a semicolons there seems wrong, as they cannot be used to separate `if` conditions and expressions; `if foo; bar` is invalid.

Maybe rewording that part of the docs and adding some one-liner examples could help?

> Instead of an indented block of code on separate lines, `then` can generally be used to define the bodies of loops and conditionals on the same line.
>   ```coffeescript
>   # Some good one-liner examples here...
>   ```

**Comment by lydell** [comment]
@epidemian :+1:

`then` already has a one-liner example in the [If, Else, Unless, and Conditional Assignment](http://coffeescript.org/#conditionals). `a = if b then c else d` is the only time I use `then` in a one-liner. The times I've used `then` with `for` is when the whole body is a `switch`. I'm not sure I've used `then` with `while`. So perhaps it is hard to come up with good examples. `while a then b` is better written as `b while a` IMO.

**Comment by connec** [comment]
Another nice use of `then` in a for loop is to run the loop body in a closure:

```coffee
for i in [1...10] then do (i) ->
  setTimeout -> console.log i
```

**Comment by loveencounterflow** [comment]
@connec yes, indeed, that looks very reasonable! Hitherto i would've written that 

```coffee
for i in [1...10] 
  do (i) ->
    setTimeout -> console.log i
```
which one might argue is more convoluted b/c of the extra level of indentation. IMHO this is a good one for the documentation ([end of section Loops and Comprehensions](http://coffeescript.org/#loops)):
```
When using a JavaScript loop to generate functions, it's common to insert a closure 
wrapper in order to ensure that loop variables are closed over, and all the generated 
functions don't just share the final values. CoffeeScript provides the `do` keyword, 
which immediately invokes a passed function, forwarding any arguments. 
[...example...]
It's possible to forgo one level of indentation by putting the `do` right after a `then` which turns the `for` statement into a one-liner as it does in a `if` ... `then` construct:
[...example...]
```

i also like that CS avoids to define the closure function multiple times:
```js
  fn = function(i) {
    return setTimeout(function() {
      return console.log(i);
    });
  };
  for (i = j = 1; j < 10; i = ++j) {
    fn(i);
  }
```
